### PR TITLE
Add a way to configure default search options

### LINKED ISSUE #4646: Remember search toggle settings, specifically Regex (or add default configuration)
### Check for existing issues

- [X] Completed

### Describe the feature

In Sublime and  Vscode, once you set "use regular expression" for searching, future searches default to having that configured.

I want to use regular expressions in search 99% of the time that I'm searching.

I would like zed to remember that I've toggled "Regex" and open _all_ search panels with that option, until I un-toggle it. If that seems unintuitive, I'd like to be able set a default value in settings.

I only really care about having Regex all the time, but I think it would make sense to have all the toggles behave the same way.

### If applicable, add mockups / screenshots to help present your vision of the feature

_No response_

### PR DESCRIPTION
**Pull request by thataboy**
Closes https://github.com/zed-industries/zed/issues/4646

```json
// Search options to enable by default when opening new project and buffer searches.
"search": {
  "whole_word": false,
  "case_sensitive": false,
  "include_ignored": false,
  "regex": false
}
```

Release Notes:

- Added `search` settings section to configure default options enabled in buffer and project searches ([#4646](https://github.com/zed-industries/zed/issues/4646))

### CODE REVIEW

**Review by SomeoneToIgnore — COMMENTED** [review]
I like the PR in general, but I think we might want to

* extract those settings into a dedicated section
* add more defaults, in particular, I crave for "Match case sensitively" default setting

Let's do both here and merge this?

File: assets/settings/default.json [hunk 1]
```diff
@@ -279,6 +279,13 @@
   "relative_line_numbers": false,
   // If 'search_wrap' is disabled, search result do not wrap around the end of the file.
   "search_wrap": true,
+  // Search options to enable by default.
+  "search_defaults": {
```
Comment (SomeoneToIgnore) [inline assets/settings/default.json:283]: What if we call it `search`?
Then we could host more settings if anything comes up in the future.

File: assets/settings/default.json [hunk 2]
```diff
@@ -279,6 +279,13 @@
   "relative_line_numbers": false,
   // If 'search_wrap' is disabled, search result do not wrap around the end of the file.
   "search_wrap": true,
+  // Search options to enable by default.
```
Comment (SomeoneToIgnore) [inline assets/settings/default.json:282]:
```suggestion
  // Search options to enable by default when opening new project and buffer searches.
```

File: crates/editor/src/editor_settings.rs [hunk 3]
```diff
@@ -155,6 +157,20 @@ pub enum ScrollBeyondLastLine {
     VerticalScrollMargin,
 }
 
+/// Default match options; all false
```
Comment (SomeoneToIgnore) [inline crates/editor/src/editor_settings.rs:160]:
```suggestion
/// Default options for buffer and project search items.
```

File: crates/editor/src/editor_settings.rs [hunk 4]
```diff
@@ -249,6 +265,11 @@ pub struct EditorSettingsContent {
     /// Default: true
     pub search_wrap: Option<bool>,
 
+    /// Defaults for search options (case sensitive, whole word, include ignored, regex)
+    ///
+    /// Default: all false
```
Comment (SomeoneToIgnore) [inline crates/editor/src/editor_settings.rs:270]:
```suggestion
    /// Default: nothing is enabled
```

File: crates/editor/src/editor_settings.rs [hunk 5]
```diff
@@ -249,6 +265,11 @@ pub struct EditorSettingsContent {
     /// Default: true
     pub search_wrap: Option<bool>,
 
+    /// Defaults for search options (case sensitive, whole word, include ignored, regex)
```
Comment (SomeoneToIgnore) [inline crates/editor/src/editor_settings.rs:268]:
```suggestion
    /// Defaults to use when opening a new buffer and project search items.
```

File: crates/search/src/buffer_search.rs [hunk 6]
```diff
@@ -504,6 +504,7 @@ impl BufferSearchBar {
         let replacement_editor = cx.new_view(|cx| Editor::single_line(cx));
         cx.subscribe(&replacement_editor, Self::on_replacement_editor_event)
             .detach();
+        let search_options = SearchOptions::from_defaults(&EditorSettings::get_global(cx).search_defaults);
```
Comment (SomeoneToIgnore) [inline crates/search/src/buffer_search.rs:507]: Unlike project searches which are instantiated on every action call, buffer search is only created a few times; once for all code-related buffers:

https://github.com/zed-industries/zed/blob/8ee42d37423ac8734595522d680aa186b617ccf3/crates/zed/src/zed.rs#L582


that's why we need to do more here, to reflect the settings changes.
I believe, we can update any of the `deploy` (or `show`/`toggle`/etc. instead?) methods to do the same settings retrieval and an update.

**Review by SomeoneToIgnore — COMMENTED** [review]
Thank you, quite nice that all both search variants are covered by the options.

Please don't get overwhelmed by the PR notes, most of them try to mention "both buffer and project searches" idea in every config-related place — feel free to use something better worded.

Apart from that, the only issue I see is the buffer search not picking up the settings changes.

File: crates/search/src/buffer_search.rs (hunk 6 above)
Comment (thataboy) [reply crates/search/src/buffer_search.rs:507]: Thanks. I'll to look to see where we can update it. You know how I can watch for a change in the settings

File: assets/settings/default.json (hunk 1 above)
Comment (thataboy) [reply assets/settings/default.json:283]: I was going to call it search_options but there's already a struct called SearchOptions. Search is a bit too wide. Maybe search_options and live with the internal and external name not matching!?

Comment (SomeoneToIgnore) [reply assets/settings/default.json:283]: I am still into just `search` as a key in that json, same as we have `git`, `terminal`, `gutter`, etc.
`"search"` seems to be vacant?

File: crates/search/src/buffer_search.rs (hunk 6 above)
Comment (SomeoneToIgnore) [reply crates/search/src/buffer_search.rs:507]: You could have another subscription like https://github.com/zed-industries/zed/blob/eecff753fbd5ab7eecc886122cd509cb45a97326/crates/outline_panel/src/outline_panel.rs#L656-L662

**Comment by thataboy** [comment]
I apologize for the merge conflict. I must be using an older or newer branch.

**Review by SomeoneToIgnore — COMMENTED** [review]
I'm a bit uncertain that resetting search toggles for existing buffer/project searches is good: imagine someone opens a project search, adds some filters and gets the results, and then keeps that project search tab somewhere in the pane for a while, opening N tabs afterwards.
Then, if we reset these, project search will retrigger and erase the old results — might not be very pleasant.

Similarly, buffer search filters changing is not very convenient if I set something custom there before altering the settings?

I think we can apply the new ones on any new project/buffer deploy after the settings change instead.

Looks good otherwise, let's fix that, the formatting and merge conflicts and we're good to go.

**Review by SomeoneToIgnore — APPROVED** [review]
Thanks, looks good!

There's a merge conflict still, so I cannot merge it yet, alas.

**Comment by thataboy** [comment]
Sorry again (I thought you could do it). I resolved the conflict but it'll no longer compile for me with that change so I'm doing it on faith. Hope it's ok.

**Comment by thataboy** [comment]
Any thoughts on how to deal with the unit test failing due to `unregistered setting type editor::editor_settings::SearchSettings`. I'm stumped. Thank you for your patience.

**Comment by SomeoneToIgnore** [comment]
I had to move the setting registration into `search::init` as it's related to both kinds of searches, and call that `search::init` after workspace initialization in every init method in the failing tests.

You've also missed a `cargo fmt` run as the CI said, that is easy to run locally before pushing.
Tests seem to pass now.

**Comment by thataboy** [comment]
Thanks. I just realized last night that if we don't want to live update the search panel, there's no need to observe settings change at all. We can just check in BufferSearchBar.show() if the settings are now different from .default_options. This would've simplified the code.

Should I do another PR or do you want to leave well enough alone? Maybe the current code could be useful in the future.

**Comment by SomeoneToIgnore** [comment]
That was my original proposal in https://github.com/zed-industries/zed/pull/17179#discussion_r1746467719 so sure, related follow-up PRs are nice.

**Comment by urandom** [comment]
This is a bit late, but shouldn't the new "use_smart_search" be part of this configuration as well? @SomeoneToIgnore

**Comment by SomeoneToIgnore** [comment]
Moving that one would be a breaking change, so it's a good thing that we did not touch it during this PR.

We have https://github.com/zed-industries/zed/issues/17742 now, so at least we should not forget to move it when the time comes.

**Pull request merged**